ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncTargetManager.java

```diff
@@ -0,0 +1,110 @@
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncState;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncTarget;
+import tech.pegasys.pantheon.ethereum.eth.sync.tasks.RetryingGetHeaderFromPeerByNumberTask;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.ethereum.p2p.wire.messages.DisconnectMessage.DisconnectReason;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+class FastSyncTargetManager<C> extends SyncTargetManager<C> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final EthContext ethContext;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+  private final BlockHeader pivotBlockHeader;
+
+  public FastSyncTargetManager(
+      final SynchronizerConfiguration config,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final SyncState syncState,
+      final LabelledMetric<OperationTimer> ethTasksTimer,
+      final BlockHeader pivotBlockHeader) {
+    super(config, protocolSchedule, protocolContext, ethContext, syncState, ethTasksTimer);
+    this.protocolSchedule = protocolSchedule;
+    this.ethContext = ethContext;
+    this.ethTasksTimer = ethTasksTimer;
+    this.pivotBlockHeader = pivotBlockHeader;
+  }
+
+  @Override
+  protected CompletableFuture<Optional<EthPeer>> selectBestAvailableSyncTarget() {
+    final Optional<EthPeer> maybeBestPeer = ethContext.getEthPeers().bestPeer();
+    if (!maybeBestPeer.isPresent()) {
+      LOG.info("No sync target, wait for peers.");
+      return completedFuture(Optional.empty());
+    } else {
+      final EthPeer bestPeer = maybeBestPeer.get();
+      if (bestPeer.chainState().getEstimatedHeight() < pivotBlockHeader.getNumber()) {
+        LOG.info("No sync target with sufficient chain height, wait for peers.");
+        return completedFuture(Optional.empty());
+      } else {
+        return confirmPivotBlockHeader(bestPeer);
+      }
+    }
+  }
+
+  private CompletableFuture<Optional<EthPeer>> confirmPivotBlockHeader(final EthPeer bestPeer) {
+    final RetryingGetHeaderFromPeerByNumberTask task =
+        RetryingGetHeaderFromPeerByNumberTask.forPivotBlock(
+            protocolSchedule,
+            ethContext,
+            ethTasksTimer,
+            pivotBlockHeader.getNumber(),
+            MAX_PIVOT_BLOCK_RETRIES);
+    task.assignPeer(bestPeer);
+    return ethContext
+        .getScheduler()
+        .timeout(task)
+        .thenApply(
+            result -> {
+              if (result.size() != 1 || !result.get(0).equals(pivotBlockHeader)) {
+                bestPeer.disconnect(DisconnectReason.USELESS_PEER);
```

Comment:
Yeah I went back and forward on it.  If we don't disconnect we'll need to maintain a sync target blacklist to avoid continuously selecting the same peer on a different chain.  But ultimately I'm ok with disconnecting because while we might be useful data from a peer it's seems like a pretty strong indicator that something is wrong with a peer that's forked 500 blocks back in disagreement with the majority of our other peers.